### PR TITLE
feat: Add support for license scanning

### LINKED ISSUE #2303: License scanner
By popular request we want Trivy to scan licenses in the scan target. This is not strictly a security scan, but it does relate to the overall risk posture of the application and also something that's commonly done as part of SCA (similar to SBOM which we already do).

Today Trivy already relates to licenses in a limited way (mostly to satisfy SBOM formats): only in container scanning, and only for some languages (where dependencies source code is available).

In line with the way other Trivy scanner work, this should show "problems" to be addressed, and not just facts about the scanned artifact. For example, vulnerability scanner doesn't show all packages by default, it shows only vulnerable packages. Similarly, license scanner doesn't show all discovered licenses by default, only potentially risky licenses (see below).

Another consideration is discovering licenses in packages vs in the filesystem. Both are interesting and serve different use-cases:
1. As a developer, I would like to verify that the software dependencies I used are legally safe in my use case. If there is an issue, I'd like to see which package was the culprit. For example, I have imported a AGPL licensed package which conflicts with the license of my application. I should either address the license requirements or remove the dependency from my app.
2. As a system operator, I would like to ensure that an artifact that I'm about to run/install is legally safe to use. If there's an issue, I'd like to be know that I should not run it, and possibly some information that can be communicated to the relevant maintainer. For example, I'm going to run some container from DockerHub, but apparently it contains a component that cannot be distributed as such. I should be alerted not to use this container, and perhaps contact the developer with technical information to seek advice.

Google had created useful framework that we can follow: https://opensource.google/documentation/reference/thirdparty/licenses
They classify third party dependencies into categories: restricted, restricted_if_statically_linked, reciprocal, notice, permissive.
They also open sourced the library used to discover licenses which also have a list of "forbidden" licenses, I presume for commercial use: https://github.com/google/licenseclassifier/blob/842c0d70d7027215932deb13801890992c9ba364/license_type.go
so I would create this ladder:

1. non-standard	
2. non-commercial (Google's forbidden)	
3. restricted	
4. restricted_if_statically_linked
5. reciprocal	
6. notice
7. permissive
8. unencumbered

And suggest to put the "risky license" threshold by default at 4 (meaning 1-4 including are bad)

to summarize:

1. Create a new scanner (`--security-check`) called `license`
2. Table for usecase#1: 
    a. License: normalized license name (in SPDX License ID format)
    b. License information: License Classification, link to `https://spdx.org/licenses/${LICENSEID}.html`
    c. Package: path to package within package manager file
    d. License and info cells should be merged in table view if possible, and hierarchically ordered in json/data formats 
    e. Example: `0BSD | unencumbered https://spdx.org/licenses/0BSD.html | pacakge.json > foo`
    e. Example: `{ "license": "0BSD", "license_classification_google": "unencumbered",  license_link: "https://spdx.org/licenses/0BSD.html", "packages": ["pacakge.json/foo"] }`
3. Table for usecase#2:
    a. Same as previous except "packages" is now "paths"
    b. since this could be potentially too verbose (e.g when every source code header contains a snippet of the license), we should group related paths that share license, to simplify the view. This can be left for a future enhancement.
4. User can tune the "risky license" threshold by classification: `--license-threshold restricted`
5. User can configure a deny-list of risky licenses which overrides `--license-threshold`: `--license-deny GPL1.0, GPL2.0`

### PR DESCRIPTION
**Pull request by owenrumney**
## Description

Add support for license scanning as an additional `--security-check license`
Scans loose source code files for licenses using go-license-detector and google-classifier as a fall back

Not configured as default scanner, needs to be run explicitly due to comparatively slower performance.

License risk threshold set to 4 by default, configurable using `--license-risk-threshold`

- Tests updated
- Added to documentation for docs site

image

## Related issues
- Close #2303 

## Checklist
- [X] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [X] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [X] I've added tests that prove my fix is effective or that my feature works.
- [X] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [X] I've added usage information (if the PR introduces new options)
- [X] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

**Comment by itaysk** [comment]
https://github.com/aquasecurity/trivy/issues/2303#issuecomment-1180635401

File: pkg/fanal/licensing/classification/classifier.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+			GoogleLicenseClassificationIndex: riskLevel,
+			GoogleLicenseClassification:      classification,
+			LicenseLink:                      licenseLink,
+		})
+	}
+
+	return license, nil
+}
+
+func GoogleClassification(licenseName string) (int, string) {
```
[the remark this answers is not in the document]
Comment (owenrumney) [reply pkg/fanal/licensing/classification/classifier.go]: @knqyf263 for my own understanding, can you clarify for me the requirement for identification and classification being separated?

[the remark this answers is not in the document]
Comment (knqyf263) [reply pkg/fanal/licensing/classification/classifier.go]: Sure. In image scanning, we basically want to leverage cache so that we can skip pulling the same layer next time. But license type classification such as "forbidden" and "restricted" can be modified by user (I'm implementing it now). In that case, we can't use the license type in cache. We need to re-classify the type according to the user's custom types. In other words, we don't have to store the license type in cache as we can't use it anyway.

1. Detect the license name such as MIT
  - It will be stored in cache
  - This information cannot be modified by user
2. Classify the license type such as "forbidden" according to the license name
  - The license name is pulled from cache
  - This type classification can be modified by user

That is why we have "analyzer" and "scanner". The analyzer stores static information in cache and a scanner performs dynamic detection based on that information in cache.

image

rather than

image

But misconfiguration and secret are a bit special because we didn't want to store file content in cache as it is large and may contain sensitive information. We also perform scanning in the analyzer.

In summary,
image

[the remark this answers is not in the document]
Comment (owenrumney) [reply pkg/fanal/licensing/classification/classifier.go]: Thanks, that makes a lot of sense. It helps solidify my understanding about the way the cache works - I had originally been passing the threshold value to the analyzer which meant once it was in the cache, threshold changes were ignored so I had to move the filtering to the report stage.

This is very helpful - thanks

**Review by knqyf263 — APPROVED** [review]

**Comment by knqyf263** [comment]
Thank you all.

**Comment by guos** [comment]
very cool function!
